translator/internal/internal_to_oc.go

```diff
@@ -0,0 +1,436 @@
+// limitations under the License.
+
+package internal
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	occommon "github.com/census-instrumentation/opencensus-proto/gen-go/agent/common/v1"
+	ocresource "github.com/census-instrumentation/opencensus-proto/gen-go/resource/v1"
+	octrace "github.com/census-instrumentation/opencensus-proto/gen-go/trace/v1"
+	"github.com/golang/protobuf/ptypes"
+	"github.com/golang/protobuf/ptypes/timestamp"
+
+	"github.com/open-telemetry/opentelemetry-collector/consumer/consumerdata"
+	"github.com/open-telemetry/opentelemetry-collector/internal"
+	"github.com/open-telemetry/opentelemetry-collector/internal/data"
+	"github.com/open-telemetry/opentelemetry-collector/translator/conventions"
+	tracetranslator "github.com/open-telemetry/opentelemetry-collector/translator/trace"
+)
+
+const sourceFormat = "ot_internal_trace"
+
+var (
+	defaultTime      = time.Time{} // zero
+	defaultTimestamp = timestamp.Timestamp{}
+	defaultProcessID = 0
+)
+
+func internalToOC(td data.TraceData) consumerdata.TraceData {
+	ocTraceData := consumerdata.TraceData{
+		SourceFormat: sourceFormat,
+	}
+
+	resourceSpansList := td.ResourceSpans()
+
+	if len(resourceSpansList) == 0 {
+		return ocTraceData
+	}
+
+	// Original resource with its spans stored in the first object of resourceSpansList
+	// Additional spans originated from different resources stored in the following objects of resourceSpansList
+	originalResourceSpans := resourceSpansList[0]
+	additionalResourceSpansList := resourceSpansList[1:]
+
+	ocTraceData.Node, ocTraceData.Resource = internalResourceToOC(originalResourceSpans.Resource())
+	ocSpans := make([]*octrace.Span, 0, td.SpanCount())
+
+	for _, span := range originalResourceSpans.Spans() {
+		ocSpans = append(ocSpans, spanToOC(span))
+	}
+	for _, resourceSpans := range additionalResourceSpansList {
+		_, resource := internalResourceToOC(resourceSpans.Resource())
+		for _, span := range resourceSpans.Spans() {
+			ocSpan := spanToOC(span)
+			ocSpan.Resource = resource
+			ocSpans = append(ocSpans, ocSpan)
+		}
+	}
+	ocTraceData.Spans = ocSpans
+
+	return ocTraceData
+}
+
+func internalResourceToOC(resource *data.Resource) (*occommon.Node, *ocresource.Resource) {
+	if resource == nil {
+		return nil, nil
+	}
+
+	attrs := resource.Attributes()
+
+	ocNode := occommon.Node{}
+	ocResource := ocresource.Resource{}
+
+	if len(attrs) == 0 {
+		return &ocNode, &ocResource
+	}
+
+	labels := make(map[string]string, len(attrs))
+	for key, attributeValue := range attrs {
+		val := attributeValueToString(attributeValue)
+
+		switch key {
+		case conventions.OCAttributeResourceType:
+			ocResource.Type = val
+		case conventions.AttributeServiceName:
+			if ocNode.ServiceInfo == nil {
+				ocNode.ServiceInfo = &occommon.ServiceInfo{}
+			}
+			ocNode.ServiceInfo.Name = val
+		case conventions.OCAttributeProcessStartTime:
+			t, err := time.Parse(time.RFC3339Nano, val)
+			if err != nil {
+				t = defaultTime
```

Comment:
I think it is probably better to leave the `StartTimestamp` field empty when it is not parseable rather than try to set it to a value.

Reply:
@tigrannajaryan doesn't this contradict you previous suggestion for OTLP translator https://github.com/open-telemetry/opentelemetry-collector/pull/582#discussion_r386160820 ?
I believe @dmitryax has adapted this code from there.
We should agree on a consistent behavior and apply it everywhere.

Reply:
Makes sense

Reply:
@tigrannajaryan doesn't this contradict you previous suggestion for OTLP translator https://github.com/open-telemetry/opentelemetry-collector/pull/582#discussion_r386160820 ?
I believe @dmitryax has adapted this code from there.
We should agree on a consistent behavior and apply it everywhere.

Reply:
Sure, let's agree on a consistent behavior. 

And another difference is reproducing of octrace.Span_TimeEvent_MessageEvent from span attributes. Not sure if it's needed fo OTLP->OC translator as well.

Reply:
@dmitryax See https://github.com/open-telemetry/opentelemetry-collector/pull/582#discussion_r385991303 re: MessageEvents. They are being passed as annotations as suggested there.
Is it different for internal -> OC? If so, why?

Reply:
@nilebox my earlier suggestion that you refer to asks to use a special value that indicates that it is an invalid time. This is not doing that. `defaultTime` is simply zero time, it does not indicate an invalid time. If we can indicate invalid time here it would be best. The problem is we can't, there are no invalid values for time possible. We used zero value in #582, but it did not achieve the goal that I had in my mind. My mistake that I did not catch it in the review of #582.

My suggestion in this review was to simply drop the StartTimestamp field completely instead of setting it to zero time. However, the difference between a missing field and zero time is negligible (missing field is slightly better since it saves space). Since we already have zero time implemented in #582, I am fine with doing the same here for consistency purposes. Works either way (either way is not ideal, but we can't do much if the data format does not allow us to indicate an error).

TLDR; ignore my rant and keep `defaultTime` as is.

Reply:
As for MessageEvents. I like the approach in this PR better. It attempts to do preserve the semantics as much as possible.

It may not matter much, but I think this is an improvement over hard-coding to Annotation only. We may want to reproduce this logic in OTLP->OC implementation or alternatively just live with it for now, since that code is temporary.